File: packages/guards/src/pii-id.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type PiiIdEntity = 'nik' | 'npwp' | 'phone-id';
interface PiiIdOptions { entities: PiiIdEntity[]; action: 'block' | 'warn' | 'mask'; }
interface PiiIdMatch { type: PiiIdEntity; value: string; start: number; end: number; }

const PATTERNS: Record<PiiIdEntity, RegExp> = {
  nik: /\b\d{16}\b/g,
  npwp: /\b\d{2}\.\d{3}\.\d{3}\.\d-\d{3}\.\d{3}\b/g,
  'phone-id': /\b(?:\+?62[-.\s]?)?0?\d{2,3}[-.\s]?\d{4}[-.\s]?\d{3,4}\b/g,
};
const MASK: Record<PiiIdEntity, string> = { nik: '[NIK]', npwp: '[NPWP]', 'phone-id': '[Telepon]' };

function detect(text: string, entities: PiiIdEntity[]): PiiIdMatch[] {
  const m: PiiIdMatch[] = [];
  for (const e of entities) { const re = new RegExp(PATTERNS[e].source, 'g'); let x; while ((x = re.exec(text))) m.push({ type: e, value: x[0], start: x.index, end: x.index + x[0].length }); }
  return m.sort((a, b) => b.start - a.start);
}

export function piiId(options: PiiIdOptions): Guard {
  return { name: 'pii-id', version: '0.1.0', description: 'Indonesian PII (NIK, NPWP, phone)', category: 'privacy', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const matches = detect(text, options.entities); const triggered = matches.length > 0;
      if (!triggered) return { guardName: 'pii-id', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      if (options.action === 'mask') { let r = text; for (const m of matches) r = r.slice(0, m.start) + MASK[m.type] + r.slice(m.end); return { guardName: 'pii-id', passed: true, action: 'override', overrideText: r, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type }) => ({ type })) } }; }
      return { guardName: 'pii-id', passed: false, action: options.action, message: `Indonesian PII: ${[...new Set(matches.map(m => m.type))].join(', ')}`, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
    },
  };
}

File: python/open_guardrail/guards/pii_id.py
"""Indonesian PII detection and masking guard."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "nik": re.compile(r"\b\d{16}\b"),
    "ktp": re.compile(r"\b\d{6}[. ]?\d{6}[. ]?\d{4}\b"),
    "npwp": re.compile(r"\b\d{2}\.\d{3}\.\d{3}\.\d-\d{3}\.\d{3}\b"),
    "passport-id": re.compile(r"\b[A-Z]\d{7,8}\b"),
    "phone-id": re.compile(r"\b(?:\+62|0062|0)\s?8\d{2}[\s\-]?\d{4}[\s\-]?\d{3,4}\b"),
}

_MASK_LABELS = {"nik": "[NIK]", "ktp": "[KTP]", "npwp": "[NPWP]", "passport-id": "[PASPOR]", "phone-id": "[TELEPON-ID]"}


def _detect(text: str, entities: list[str]) -> list[dict]:
    matches = []
    for e in entities:
        pat = _PATTERNS.get(e)
        if pat:
            for m in pat.finditer(text):
                matches.append({"type": e, "value": m.group(), "start": m.start(), "end": m.end()})
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask(text: str, matches: list[dict]) -> str:
    r = text
    for m in matches:
        r = r[: m["start"]] + _MASK_LABELS.get(m["type"], "[PII]") + r[m["end"] :]
    return r


class _PiiId:
    def __init__(self, *, action: str = "block", entities: Optional[List[str]] = None) -> None:
        self.name = "pii-id"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000
        if not matches:
            return GuardResult(guard_name="pii-id", passed=True, action="allow", latency_ms=round(elapsed, 2))
        detected = [{"type": m["type"], "value": m["value"]} for m in matches]
        if self.action == "mask":
            return GuardResult(guard_name="pii-id", passed=True, action="override", override_text=_mask(text, matches), latency_ms=round(elapsed, 2), details={"detected": detected})
        return GuardResult(guard_name="pii-id", passed=False, action=self.action, message="Indonesian PII detected", latency_ms=round(elapsed, 2), details={"detected": detected})


def pii_id(*, action: str = "block", entities: Optional[List[str]] = None) -> _PiiId:
    return _PiiId(action=action, entities=entities)
